Added a more user friendly ImageReader

andrewkaufman opened the pull request:
Renamed ImageReader to OpenImageIOReader, and created a new ImageReader which embeds the OpenImageIOReader along with a ColorSpace conversion to linear.

This isn't ready for review, as I'm still missing the error handling and range holding features we've discussed, but I'm putting it up now to see if a jpg comparison test that fails at IE actually passes on Travis.

andrewkaufman commented:
I think this is ready for review now (pending the tests passing), with all the features we've discussed. I did name a few things a bit differently to what we had discussed, because I found the names a bit misleading based on what actually happens in the implementation.

For example, "nearest frame" error mode was renamed to "Hold" because it doesn't do any look-ahead as "nearest" might imply, it always holds the previous valid frame of the sequence.

The frame range stuff I decided to call a mask (frameRangeMask and frameRangeMaskMode) because it seemed more accurate than just calling it frameRange. Unfortunately I couldn't have  a "Black" mode for the range mask because of silly c++ non-namespaced enums... so I named it "BlackOutside" which I'm not sure if I really like... Same goes for "Hold" which I named "ClampToRange", though that one I do like a bit more I think.

lucienfostier commented on src/GafferImage/OpenImageIOReader.cpp:
```diff
@@ -0,0 +1,625 @@
+					std::vector<IECore::FrameList::Frame> frames;
+					sequence->getFrameList()->asList( frames );
+					IECore::FrameList::Frame currentFrame = sequence->frameForFileName( fileNameStr );
+					std::vector<IECore::FrameList::Frame>::iterator fIt = std::lower_bound( frames.begin(), frames.end(), currentFrame );
```
@andrewkaufman
what happen if we dont have a valid first frame?

andrewkaufman replied:
You get an error, because if there's no valid first frame, then this isn't a valid image sequence.

andrewkaufman replied:
I guess you're wondering if it should fall back to the default format with an all black image instead of erroring?

lucienfostier replied:
i ask that because we discuss it with lorenzo and in IEMedia Ehsan did a sustem to handle such a case.

If you dont have a valid first (expected) frame, we look for the first valid frame.

So say you expect a range starting at 1001 but the frame sequence your dealing with start at 1005, you hold 1005 during 1001 to 1004.

I ask you the question because Im not sure if this node has a concept of expected frame range, maybe it's more like, I deal with what I have and handle error that may rise in the frame range.



> On Nov 6, 2015, at 14:14, Andrew Kaufman <notifications@github.com> wrote:
> 
> In src/GafferImage/OpenImageIOReader.cpp:
> 
> > +			IECore::ls( dir.string(), sequences, /* minSequenceSize */ 1 );
> > +			for( std::vector<IECore::FileSequencePtr>::iterator it = sequences.begin(); it != sequences.end(); ++it )
> > +			{
> > +				IECore::FileSequence *sequence = it->get();
> > +				std::string prefix = ( dir / boost::filesystem::path( sequence->getPrefix() ) ).string();
> > +				sequence->setPrefix( prefix );
> > +				std::string suffix = sequence->getSuffix();
> > +				if (
> > +					( fileNameStr.substr( 0, prefix.size() ) == prefix ) &&
> > +					( fileNameStr.substr( fileNameStr.size() - suffix.size(), suffix.size() ) == suffix )
> > +				)
> > +				{
> > +					std::vector<IECore::FrameList::Frame> frames;
> > +					sequence->getFrameList()->asList( frames );
> > +					IECore::FrameList::Frame currentFrame = sequence->frameForFileName( fileNameStr );
> > +					std::vector<IECore::FrameList::Frame>::iterator fIt = std::lower_bound( frames.begin(), frames.end(), currentFrame );
> I guess you're wondering if it should fall back to the default format with an all black image instead of erroring?
> 
> —
> Reply to this email directly or view it on GitHub.
>

andrewkaufman replied:
Yeah, this node is the lower level one that like you say "deals with what I have and handle error that may rise in the frame range".

But the higher level one in the next commits on this PR does have a frameRangeMask, and in that case you would set the expected start frame to 1005 and the mode to ClampToRange. But I guess maybe this highlights an issue with that expectation, which is in this case, the user doesn't necessarily want to set an expected start frame, but rather wants some auto-magic to find one for them?

andrewkaufman commented:
After running a test for my Daily system, I ran into the problem Lucien was describing straight away. Luckily the diff was quite minimal (and it actually simplified some funky logic in the test cases) so I decided to rebase it in.

This change probably does bring my choice of naming into question a bit more though. Despite the extra logic, I still don't think ErrorMode "Hold" can be considered "Nearest", but its not really "Hold" anymore either. Now its more like "HoldPreviousUnlessThereIsNoPreviousInWhichCaseClampToRange" but that obviously won't do. Any thoughts on better enum names/values? Or should I just stick with what I've got?

andrewkaufman commented:
I've addressed a lot of the comments, but unfortunately I rebased on master and that seems to have wiped them all out... they're still findable via email links to old commits.

Luckily I kept track for the most part with my individual commits (with a plan to rebase once we're all done), so you can see most of what I've addressed. New commits start with 7769e82.

I think what remains is:

- dropping the UI commits @johnhaddon didn't like, and adding a custom UI for frameStartMask and frameEndMask (see https://github.com/andrewkaufman/gaffer/commit/3267ddd021baa4b574e259700fd1f9227b786df9#commitcomment-14284539)
- `Context::Scope` handling NULL contexts (see https://github.com/andrewkaufman/gaffer/commit/3267ddd021baa4b574e259700fd1f9227b786df9#commitcomment-14283613)
- The ColorSpace / GammaCorrected discussion (see https://github.com/andrewkaufman/gaffer/commit/3267ddd021baa4b574e259700fd1f9227b786df9#commitcomment-14283515)
- The ErrorMode::Hold file search discussion (see https://github.com/andrewkaufman/gaffer/commit/8c6914c3685ab02a1641558f166dc3465df3aeb0#commitcomment-14287527)

andrewkaufman commented:
Ok, I believe I've addressed all the discussion points now. I decided to squash it all down now, since the final code was quite different from the original commits anyways.

andrewkaufman commented:
I'm not sure about this failing test... it is using an ImageReader and when I swap it for an OpenImageIOReader it passes, so it looks like I've done something wrong. Can anyone spot it? Or will merging https://github.com/ImageEngine/gaffer/pull/1524 positively affect this?

```
======================================================================

FAIL: testOutOfBoundsSampleModeBlack (GafferImageTest.SamplerTest.SamplerTest)

----------------------------------------------------------------------

Traceback (most recent call last):

File "/home/travis/build/ImageEngine/gaffer/install/gaffer-0.18.0.0-linux/python/GafferImageTest/SamplerTest.py", line 104, in testOutOfBoundsSampleModeBlack

self.assertNotEqual( s.sample( bounds.min.x+.5, bounds.min.y+.5 ), 0. )

AssertionError: 0.0 == 0.0

----------------------------------------------------------------------
```

andrewkaufman commented:
Indeed when I merge #1524 on top of this, that test passes....

johnhaddon commented:
Hmm, that's pretty odd isn't it? If it passes for an OpenImageIOReader, it really should pass with an ImageReader too I would have thought - do we know what the difference is between the two in this scenario?

andrewkaufman commented:
The differences I could think of are that ImageReader has an extra ColorSpace node (or the default value) feeding its out image, and that OpenImageIOReader['out'] is flagged non-cachable.

andrewkaufman commented:
I monkeyed with the failing test a bit to see if intermediate nodes work, and it seems using both `r['__oiioReader']['out']` and `r['__colorSpace']['out']` give the expected result, but using `r['__intermediateImage']` doesn't. What seems to be happening is that `m_dataWindow` computed in `Sampler::init` is receiving an un-initialized dataWindow `(2147483647 2147483647) to (-2147483648 -2147483648)` for `r['__intermediateImage']` rather than the expected dataWindow `(0 0) to (200 150)`

What's puzzling is that if I simply call `r["__intermediateImage"]["dataWindow"].getValue()` I do get the expected value. So for some reason, the `m_plug->dataWindowPlug()->getValue()` call in `Sampler::init` is getting this other value...

andrewkaufman commented:
Looking closer at the commits in #1524 though, the test passes because it no longer uses an ImageReader... so that's probably not the best gauge for a fix...

andrewkaufman commented:
It seems like making a Sampler at all is somehow messing with the result of `r["__intermediateImage"]["dataWindow"].getValue()` though I don't understand why...

```
print "__intermediateImage", r["__intermediateImage"]["dataWindow"].getValue()
s = GafferImage.Sampler( r["__oiioReader"]["out"], "R", bounds, f, GafferImage.Sampler.BoundingMode.Black )
print "__intermediateImage", r["__intermediateImage"]["dataWindow"].getValue()
```

gives

```
__intermediateImage 0 0 200 150
__intermediateImage 2147483647 2147483647 -2147483648 -2147483648
```

johnhaddon commented on src/GafferImage/OpenImageIOReader.cpp:
```diff
@@ -0,0 +1,699 @@
+	// but we internally vary the result output by
+	// frame, we need to explicitly hash the frame
+	// when the value contains FrameSubstitutions.
+	const std::string &fileName = fileNamePlug()->getValue();
```
I'm surprised the compiler is allowing this, since `getValue()` returns a `string` rather than `string &`. I don't see anything to stop the string being destroyed before it gets hashed.

johnhaddon commented:
That behaviour you're seeing is mighty odd - my best guess would be that a hashing bug is causing the wrong values to be pulled from the cache, but that's just a guess. Best thing might be to instrument things a bit to see where the value is coming from...

andrewkaufman commented:
I've rebased on master and added cdfbb36 which changes the mask plug naming based on a discussion with @johnhaddon
